### PR TITLE
CNV-32097: add column management to the custom list filter

### PR DESCRIPTION
**Pull request by upalatucci**
## 📝 Description

Add Column Management modal to the custom ListPageFilter component

## 🎥 Demo

https://github.com/kubevirt-ui/kubevirt-plugin/assets/29160323/41c216a2-796e-49f2-a12d-5b33dc64ddc6

### CODE REVIEW

File: src/utils/components/ColumnManagementModal/ColumnManagementModal.tsx [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+  const defaultColumns = columnLayout.columns.filter((column) => column.id && !column.additional);
+  const additionalColumns = columnLayout.columns.filter((column) => column.additional);
+
+  const [_, setActiveColumns, loaded, error] = useKubevirtUserSettingsTableColumns({
+    columnManagementID: columnLayout.id,
+    columns: columnLayout.columns,
+  });
+
+  const [checkedColumns, setCheckedColumns] = React.useState<Set<string>>(
+    columnLayout.selectedColumns && columnLayout.selectedColumns.size !== 0
```
Comment (avivtur) [inline src/utils/components/ColumnManagementModal/ColumnManagementModal.tsx:47]: can you please check if `!isEmpty(columnLayout.selectedColumns) ? .... : .... ` would work here?

File: src/utils/components/ColumnManagementModal/ColumnManagementModal.tsx [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+      : updatedCheckedColumns.add(selectedId);
+    setCheckedColumns(updatedCheckedColumns);
+  };
+
+  const submit: MouseEventHandler<HTMLButtonElement> = async (event) => {
+    event.preventDefault();
+    const orderedCheckedColumns = new Set<string>();
+    columnLayout.columns.forEach(
+      (column) => checkedColumns.has(column.id) && orderedCheckedColumns.add(column.id),
+    );
```
Comment (avivtur) [inline src/utils/components/ColumnManagementModal/ColumnManagementModal.tsx:66]: Can you please check if checkedColumes.forEach((col) => orderedCheckedColumns.add(col.id)) works the same?

File: src/utils/components/ColumnManagementModal/ColumnManagementModal.tsx [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+    await setActiveColumns([...orderedCheckedColumns]);
+    onClose();
+  };
+
+  const areMaxColumnsDisplayed = checkedColumns.size >= MAX_VIEW_COLS;
+
+  const resetColumns = (event: React.SyntheticEvent): void => {
```
Comment (avivtur) [inline src/utils/components/ColumnManagementModal/ColumnManagementModal.tsx:74]: Here too please remove React.

File: src/utils/components/ColumnManagementModal/DataListRow.tsx [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+type DataListRowProps = {
+  checkedColumns: Set<string>;
+  column: ManagedColumn;
+  disableUncheckedRow: boolean;
+  inputId: string;
+  onChange: (checked: boolean, event: React.SyntheticEvent) => void;
+};
+
+const DataListRow: React.FC<DataListRowProps> = ({
```
Comment (avivtur) [inline src/utils/components/ColumnManagementModal/DataListRow.tsx:22]: Here too please drop React.

File: src/utils/components/ListPageFilter/ListPageFilter.tsx [hunk 5]
```diff
@@ -186,6 +204,26 @@ const ListPageFilter: FC<ListPageFilterProps> = ({ data, loaded, onFilterChange,
             </ToolbarFilter>
           </ToolbarItem>
         </ToolbarToggleGroup>
+        {columnLayout?.id && !hideColumnManagement && (
```
Comment (avivtur) [inline src/utils/components/ListPageFilter/ListPageFilter.tsx:207]: I'd move this component to a seprate file as this file is already over 200 lines

**Review by avivtur — COMMENTED** [review]
Hi Ugo excellent work! 
I left a few comments, And I wonder about something:
The component we get from SDK now shows the namespace checkbox when we are scoped under any namespace, and it is not really functional, can you disable this checkbox when the current namespace exists (=not all namespaces)?

File: src/utils/components/ColumnManagementModal/ColumnManagementModal.tsx (hunk 1 above)
Comment (upalatucci) [reply src/utils/components/ColumnManagementModal/ColumnManagementModal.tsx:47]: but `isEmpty` will not check the size.

File: src/utils/components/ColumnManagementModal/ColumnManagementModal.tsx [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+};
+
+export const ColumnManagementModal: FC<ColumnManagementModalProps> = ({
+  columnLayout,
+  isOpen,
+  onClose,
+}) => {
+  const { t } = useKubevirtTranslation();
+  const defaultColumns = columnLayout.columns.filter((column) => column.id && !column.additional);
+  const additionalColumns = columnLayout.columns.filter((column) => column.additional);
```
[the remark this answers is not in the document]
Comment (upalatucci) [reply src/utils/components/ColumnManagementModal/ColumnManagementModal.tsx:39]: I think the filtering is more explanatory of what we are doing. but thanks :-)

File: src/utils/components/ColumnManagementModal/ColumnManagementModal.tsx [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+
+  const [checkedColumns, setCheckedColumns] = React.useState<Set<string>>(
+    columnLayout.selectedColumns && columnLayout.selectedColumns.size !== 0
+      ? new Set(columnLayout.selectedColumns)
+      : new Set(defaultColumns.map((col) => col.id)),
+  );
+
+  const onColumnChange = (checked: boolean, event: React.SyntheticEvent): void => {
+    const updatedCheckedColumns = new Set<string>(checkedColumns);
+    const selectedId = event?.currentTarget?.id?.slice(4);
```
[the remark this answers is not in the document]
Comment (upalatucci) [reply src/utils/components/ColumnManagementModal/ColumnManagementModal.tsx:54]: I'll create a function to be more explanatory

File: src/utils/components/ColumnManagementModal/ColumnManagementModal.tsx (hunk 1 above)
Comment (avivtur) [reply src/utils/components/ColumnManagementModal/ColumnManagementModal.tsx]: Have you looked at the definition of isEmpty?
image

it should also check the length to make sure if it's empty or not.
I've made also a little check in codesandbox: https://codesandbox.io/s/wizardly-bassi-xxwg9s?file=/script.js

**Comment by avivtur** [comment]
/lgtm
Perfect! Thanks @upalatucci :100:

File: src/utils/components/ColumnManagementModal/ColumnManagementModal.tsx (hunk 1 above)
Comment (avivtur) [reply src/utils/components/ColumnManagementModal/ColumnManagementModal.tsx]: Please look at isEmpty definition, you'll see that it does check size as well

**Comment by avivtur** [comment]
/hold 
@upalatucci please check why tests are failing

**Pull request merged**